File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.cc
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int N, M, R;
    cin >> N >> M >> R;

    vector<vector<int>> adj(N + 1);
    vector<int> visited(N + 1, 0);
    int cnt = 1;

    for (int i = 0; i < M; i++) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    for (int i = 1; i <= N; i++) {
        sort(adj[i].begin(), adj[i].end());
    }

    queue<int> q;
    q.push(R);
    visited[R] = cnt;

    while (!q.empty()) {
        int num = q.front();
        q.pop();
        for (int i : adj[num]) {
            if (visited[i] == 0) {
                cnt++;
                visited[i] = cnt;
                q.push(i);
            }
        }
    }

    for (int i = 1; i <= N; i++) {
        cout << visited[i] << '\n';
    }

    return 0;
}

File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.py
from collections import deque
import sys
input = sys.stdin.readline

N, M, R = map(int, input().split(' '))

adj = [[] for i in range(N+1)]

for i in range(M):
    u, v = map(int, input().split(' '))
    adj[u].append(v)
    adj[v].append(u)


visited = [0] * (N+1)
deq = deque([])
cnt = 1

def bfs(R):
    deq.append(R)
    global cnt
    visited[R] = cnt
    while deq:
        num = deq.popleft()
        adj[num].sort()
        for i in adj[num]:
            if not visited[i]:
                cnt += 1
                visited[i] = cnt
                deq.append(i)

bfs(R)

for i in range(1, len(visited)):
    print(visited[i])